### PR TITLE
build in a docker environment

### PR DESCRIPTION
**Pull request by gregwebs**
This PR add a docker build. You can read more about it in the documentation file in the PR.

I was having trouble building this project so I switched to a dockerized build and thought I would share it with everyone.
This could greatly improve things for someone looking to make a first contribution or those looking to build from source to improve providence.

### CODE REVIEW

**Comment by chinggg** [comment]
Thanks! It sounds like a good idea to have a Docker setup. I just realize that our official repo does not contain any Dockerfile or CI/CD file for building bitcoin. Though Cirrus CI is always running, it seems to check things other than the build process. FYI, I can find a Dockerfile in https://github.com/google/oss-fuzz/blob/master/projects/bitcoin-core/Dockerfile.

**Comment by sipa** [comment]
There has been earlier discussion here: https://github.com/bitcoin-core/packaging/issues/55#issuecomment-1125393507

**Comment by gregwebs** [comment]
There is an image from that issue that points to a good [Dockerfile](https://github.com/lncm/docker-bitcoind). Although I saw that Dockerfile and it has lots of things and I didn't understand how they all worked.

But that is for packaging a reproducible build. This Dockerfile is designed for a development workflow. Those are two different use cases.

**Comment by MarcoFalke** [comment]
> But that is for packaging a reproducible build. This Dockerfile is designed for a development workflow. Those are two different use cases.

What developers are you referring to? People working on Bitcoin Core that use docker/podman/whatever already have their scripts for their use cases fledged out. I don't think the dockerfile in this pull is going to supersede any of that. It seems more like a dev environment for your use case.

Some more references about docker use:

* For dev with depends: https://github.com/fanquake/core-review/tree/75d8705bfcf0826a1534d6dc25e8fea61b9a0113/docker
* As pointed out by @chinggg the OSS-Fuzz Dockerfile
* The CI docker use: https://github.com/bitcoin/bitcoin/blob/194f6dc43ccc330a8a4607be3a2b8935490d6db0/ci/test/04_install.sh#L44
* Probably a lot more

So I am wondering for whom you are writing the dockerfile in this pull.

**Comment by chinggg** [comment]
> So I am wondering for whom you are writing the dockerfile in this pull.

@MarcoFalke I think he is writing Dockerfile for someone who is new to Bitcoin-Core and has problem building it? As he said at first:
> This could greatly improve things for someone looking to make a first contribution or those looking to build from source to improve providence.

However, I am also new to bitcoin but I think `build-unix.md` already teach me how to build, which is basically installing dependencies then `configure` and `make`. Personally I won't use Docker if I am new to a project.

**Comment by MarcoFalke** [comment]
See also https://github.com/bitcoin/bitcoin/pull/10903 (Add configuration files for a Docker-based development environment)

File: build-docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,32 @@
+FROM alpine
+
+# build depndencies
+RUN apk update && apk add \
```
Comment (jarolrod) [inline build-docker/Dockerfile:4]: This dependency list is making a biased decision on how to build bitcoin core, primarily on not including qt related dependencies to build `bitcoin-qt`

**Review by jarolrod — COMMENTED** [review]
concept nack 

I don't see a strong motivation for introducing this into the repo. Ideally, someone completely new to the project could quickly get up to speed with the build docs and be able to successfully compile on their own. With the docs across `build-*` and `depends/`, I'd assume that trouble compiling is an issue of "not reading the docs" or us not placing information in the best way/correct places.

> I was having trouble building this project  ...

Any feedback on how we can improve the build documentation would be great.

**Comment by MarcoFalke** [comment]
Concept ACK, if this helps on-board new devs or improves UX for existing devs, but it would be good to hear from at least one dev that will be using this first to avoid adding an unused script.

**Comment by laanwj** [comment]
I have nothing against including a docker config, but it should not be required. I am sometimes annoyed by projects that have docker builds as the only (realistic) option for doing development. And it's not like we don't already have the `depends` system to facilitate easy install of dependencies :smile: .

**Comment by fanquake** [comment]
Concept NACK. This isn't something we need to maintain in this repo. If anything, maybe the discussion can be restarted in https://github.com/bitcoin-core/packaging.

> I was having trouble building this project

What issue were you having? Why was introducing Docker the solution?

File: build-docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  make \
+  file \
+  autoconf \
+  automake \
+  build-base \
+  libtool \
+  db-c++ \
+  db-dev \
+  boost-system \
+  boost-program_options \
```
Comment (fanquake) [inline build-docker/Dockerfile:15]: This list includes a number of dependencies that are not required to compile Bitcoin Core. i.e boost-filesystem, boost-system, boost-program-options, libressl.

**Comment by gregwebs** [comment]
Personally I don’t chase dependencies for snowflake installs anymore. I use Docker, nix, or a shell script and a single install command.

It seems there is a desire for a docker build given all the things that are being pointed to from this pull request.

With that being said I would rather not convince maintainers to maintain something they don’t want to. I should be able to put this into a separate repo. Then I think what is missing is some documentation suggesting these community solutions for those wanting to use Docker.

File: build-docker/Dockerfile (hunk 2 above)
Comment (gregwebs) [reply build-docker/Dockerfile]: I removed boost-program-options. Other docker images that the community has created contain the rest of these libraries so it seems like they are needed. It is now easy to test this out with the docker build, but I am not familiar with all the configure options.

**Comment by luke-jr** [comment]
This seems fine if there are developers (like @gregwebs) willing to maintain it. At least 2 others should probably agree to be available for reviewing it.

Mergers are supposed to be janitorial, not dictators over what is supported. If there are developers willing to do the work, that should be sufficient.

(Note that I do not personally use Docker, and will not be reviewing Docker stuff)

**Comment by gregwebs** [comment]
@luke-jr thank you for the support. However, after making a separate repo I like it that way. I think it is just as convenient as in repo, and in some cases actually better because it is easy to build an older version. The reason to put it in repo is for it to be more official and to try to ensure a better review process. However, a Dockerfile is easy to audit, and the demand for this approach is unknown right now.

I would still like to create a document linking to different community efforts at Docker builds.

**Pull request closed without merging**